File: Problems/P0054/experiments/g4.12-exact-fifty-four/middle-seven-nine-extension.py
#!/usr/bin/env python3
"""Test the four-row extensions in the G4.12 (7,9) split."""

from __future__ import annotations

import argparse
import hashlib
import itertools
import json
from collections.abc import Iterator
from pathlib import Path


def sha256_file(path: Path) -> str:
    digest = hashlib.sha256()
    with path.open("rb") as handle:
        for chunk in iter(lambda: handle.read(1 << 20), b""):
            digest.update(chunk)
    return digest.hexdigest()


def minimize(rows: set[frozenset[int]]) -> tuple[frozenset[int], ...]:
    result: list[frozenset[int]] = []
    for row in sorted(rows, key=lambda current: (len(current), tuple(sorted(current)))):
        if not any(previous <= row for previous in result):
            result.append(row)
    return tuple(result)


def blockers(rows: tuple[frozenset[int], ...]) -> tuple[frozenset[int], ...]:
    candidates: tuple[frozenset[int], ...] = (frozenset(),)
    for row in rows:
        expanded: set[frozenset[int]] = set()
        for candidate in candidates:
            if candidate & row:
                expanded.add(candidate)
            else:
                expanded.update(candidate | {point} for point in row)
        candidates = minimize(expanded)
    return candidates


def is_clutter(rows: tuple[frozenset[int], ...]) -> bool:
    return all(
        left == right or not rows[left] <= rows[right]
        for left in range(len(rows))
        for right in range(len(rows))
    )


def point_degrees(rows: tuple[frozenset[int], ...]) -> dict[int, int]:
    points = frozenset().union(*rows)
    return {point: sum(point in row for row in rows) for point in points}


def parse_classes(path: Path) -> tuple[str, ...]:
    payload = json.loads(path.read_text(encoding="ascii"))
    classes = payload.get("classes")
    if not isinstance(classes, list) or not all(isinstance(key, str) for key in classes):
        raise TypeError("invalid JSON kernel classes")
    return tuple(classes)


def kernel_from_key(key: str) -> tuple[tuple[frozenset[int], ...], int]:
    supports = []
    support_key = key.split("|", 1)[0]
    for item in support_key.rstrip(",").split(","):
        support, value = item.split(":")
        supports.extend([int(support)] * int(value))
    rows = tuple(
        frozenset(point for point, support in enumerate(supports) if support & (1 << row))
        for row in range(4)
    )
    return rows, len(supports)


def single_row_candidates(
    base: tuple[frozenset[int], ...],
    x: int,
) -> Iterator[tuple[frozenset[int], ...]]:
    degrees = point_degrees(base)
    eligible = tuple(point for point in range(x + 1) if degrees.get(point, 0) < 4)
    for size in range(len(eligible) + 1):
        for chosen in itertools.combinations(eligible, size):
            for fresh in range(8 - size):
                if not 3 <= size + fresh <= 7:
                    continue
                row = frozenset(chosen) | frozenset(range(x + 1, x + 1 + fresh))
                yield (row,)


def two_row_candidates(
    base: tuple[frozenset[int], ...],
    x: int,
) -> Iterator[tuple[frozenset[int], ...]]:
    degrees = point_degrees(base)
    existing = tuple(range(x + 1))
    for patterns in itertools.product(range(4), repeat=len(existing)):
        if any(
            degrees.get(point, 0) + bool(pattern & 1) + bool(pattern & 2) > 4
            for point, pattern in zip(existing, patterns, strict=True)
        ):
            continue
        first = frozenset(
            point for point, pattern in zip(existing, patterns, strict=True) if pattern & 1
        )
        second = frozenset(
            point for point, pattern in zip(existing, patterns, strict=True) if pattern & 2
        )
        for first_private in range(8 - len(first)):
            for second_private in range(8 - len(second)):
                max_shared = min(
                    7 - len(first) - first_private,
                    7 - len(second) - second_private,
                )
                for shared in range(max_shared + 1):
                    if not 3 <= len(first) + first_private + shared <= 7:
                        continue
                    if not 3 <= len(second) + second_private + shared <= 7:
                        continue
                    next_point = x + 1
                    first_fresh = frozenset(range(next_point, next_point + first_private))
                    next_point += first_private
                    second_fresh = frozenset(range(next_point, next_point + second_private))
                    next_point += second_private
                    common_fresh = frozenset(range(next_point, next_point + shared))
                    rows = (
                        first | first_fresh | common_fresh,
                        second | second_fresh | common_fresh,
                    )
                    if tuple(sorted(rows, key=lambda row: tuple(sorted(row)))) != rows:
                        continue
                    yield rows


def valid_extension(
    rows: tuple[frozenset[int], ...],
    kernel: tuple[frozenset[int], ...],
    x: int,
) -> bool:
    if len(rows) != 9 or not is_clutter(rows):
        return False
    if any(not 3 <= len(row) <= 7 for row in rows):
        return False
    if max(point_degrees(rows).values()) > 4:
        return False
    dual = blockers(rows)
    if len(dual) != 7 or not all(row in dual for row in kernel):
        return False
    if any(not 3 <= len(row) <= 9 for row in dual):
        return False
    if max(point_degrees(dual).values()) > 3:
        return False
    return sum(x in row for row in dual) == 3


def check_class(key: str) -> dict[str, int | str]:
    kernel, old_points = kernel_from_key(key)
    kernel_blockers = blockers(kernel)
    x = old_points
    pair_blockers = tuple(row for row in kernel_blockers if len(row) == 2)
    other_blockers = tuple(row for row in kernel_blockers if len(row) != 2)
    if len(pair_blockers) not in (3, 4):
        raise AssertionError("unexpected pair-blocker count")

    optional_count = 4 - len(pair_blockers)
    added_options = tuple(
        added
        for size in range(optional_count + 1)
        for added in itertools.combinations(range(len(other_blockers)), size)
    )
    candidate_count = 0
    valid = 0
    for added in added_options:
        base = tuple(row | {x} for row in pair_blockers) + tuple(
            row | {x} if index in added else row for index, row in enumerate(other_blockers)
        )
        extras_needed = 9 - len(base)
        if extras_needed == 0:
            candidates: Iterator[tuple[frozenset[int], ...]] = iter(((),))
        elif extras_needed == 1:
            candidates = single_row_candidates(base, x)
        elif extras_needed == 2:
            candidates = two_row_candidates(base, x)
        else:
            raise AssertionError("unexpected kernel blocker count")
        for extras in candidates:
            rows = base + extras
            if not is_clutter(rows):
                continue
            candidate_count += 1
            valid += valid_extension(rows, kernel, x)
    return {
        "kernel": key,
        "points": old_points,
        "kernel_blockers": len(kernel_blockers),
        "pair_blockers": len(pair_blockers),
        "extension_candidates": candidate_count,
        "valid_extensions": valid,
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--kernels", type=Path, required=True)
    parser.add_argument("--output", type=Path, required=True)
    args = parser.parse_args()

    results = [check_class(key) for key in parse_classes(args.kernels)]
    report = {
        "schema": "p0054.g4.12.middle-seven-nine-extension.v1",
        "epistemic_label": "COMPUTED",
        "kernel_path": str(args.kernels),
        "kernel_sha256": sha256_file(args.kernels),
        "python_source_sha256": sha256_file(Path(__file__)),
        "kernel_classes": len(results),
        "extension_candidates": sum(int(result["extension_candidates"]) for result in results),
        "valid_extensions": sum(int(result["valid_extensions"]) for result in results),
        "classes": results,
    }
    args.output.write_text(json.dumps(report, indent=2, sort_keys=True) + "\n", encoding="ascii")
    print(json.dumps(report, indent=2, sort_keys=True))


if __name__ == "__main__":
    main()

File: Problems/P0054/experiments/g4.12-exact-fifty-four/middle-seven-nine-extension.cpp
#include <algorithm>
#include <bit>
#include <cstdint>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <string>
#include <utility>
#include <vector>

using Mask = std::uint64_t;

struct Summary {
	std::uint64_t candidates = 0;
	std::uint64_t valid = 0;
};

int size(Mask mask)
{
	return std::popcount(mask);
}

std::vector<Mask> minimize(std::vector<Mask> rows)
{
	std::sort(rows.begin(), rows.end(), [](Mask left, Mask right) {
		return std::make_pair(size(left), left) < std::make_pair(size(right), right);
	});
	rows.erase(std::unique(rows.begin(), rows.end()), rows.end());
	std::vector<Mask> result;
	for (Mask row : rows) {
		bool dominated = false;
		for (Mask previous : result) {
			if ((previous & row) == previous) {
				dominated = true;
				break;
			}
		}
		if (!dominated) result.push_back(row);
	}
	return result;
}

std::vector<Mask> blockers(const std::vector<Mask> &rows)
{
	std::vector<Mask> candidates{0};
	for (Mask row : rows) {
		std::vector<Mask> expanded;
		for (Mask candidate : candidates) {
			if (candidate & row) {
				expanded.push_back(candidate);
				continue;
			}
			for (int point = 0; point < 64; ++point) {
				if (row & (Mask{1} << point)) {
					expanded.push_back(candidate | (Mask{1} << point));
				}
			}
		}
		candidates = minimize(std::move(expanded));
	}
	return candidates;
}

bool is_clutter(const std::vector<Mask> &rows)
{
	for (int left = 0; left < static_cast<int>(rows.size()); ++left) {
		for (int right = 0; right < static_cast<int>(rows.size()); ++right) {
			if (left != right && (rows[left] & rows[right]) == rows[left]) return false;
		}
	}
	return true;
}

std::vector<int> degrees(const std::vector<Mask> &rows, int point_count)
{
	std::vector<int> result(point_count);
	for (Mask row : rows) {
		for (int point = 0; point < point_count; ++point) {
			if (row & (Mask{1} << point)) ++result[point];
		}
	}
	return result;
}

bool contains_all(const std::vector<Mask> &rows, const std::vector<Mask> &kernel)
{
	for (Mask row : kernel) {
		if (std::find(rows.begin(), rows.end(), row) == rows.end()) return false;
	}
	return true;
}

std::vector<int> points(Mask row)
{
	std::vector<int> result;
	for (int point = 0; point < 64; ++point) {
		if (row & (Mask{1} << point)) result.push_back(point);
	}
	return result;
}

bool valid_extension(
	const std::vector<Mask> &rows,
	const std::vector<Mask> &kernel,
	int point_count,
	int x
)
{
	if (rows.size() != 9 || !is_clutter(rows)) return false;
	for (Mask row : rows) {
		if (size(row) < 3 || size(row) > 7) return false;
	}
	const std::vector<int> row_degrees = degrees(rows, point_count);
	if (*std::max_element(row_degrees.begin(), row_degrees.end()) > 4) return false;
	const std::vector<Mask> dual = blockers(rows);
	if (dual.size() != 7 || !contains_all(dual, kernel)) return false;
	for (Mask row : dual) {
		if (size(row) < 3 || size(row) > 9) return false;
	}
	const std::vector<int> dual_degrees = degrees(dual, point_count);
	if (*std::max_element(dual_degrees.begin(), dual_degrees.end()) > 3) return false;
	return dual_degrees[x] == 3;
}

void test_candidate(
	const std::vector<Mask> &rows,
	const std::vector<Mask> &kernel,
	int point_count,
	int x,
	Summary &summary
)
{
	if (!is_clutter(rows)) return;
	++summary.candidates;
	if (valid_extension(rows, kernel, point_count, x)) ++summary.valid;
}

void enumerate_one_row(
	const std::vector<Mask> &base,
	const std::vector<Mask> &kernel,
	int x,
	Summary &summary
)
{
	const std::vector<int> base_degrees = degrees(base, x + 1);
	std::vector<int> eligible;
	for (int point = 0; point <= x; ++point) {
		if (base_degrees[point] < 4) eligible.push_back(point);
	}
	for (std::uint64_t chosen = 0;
		chosen < (std::uint64_t{1} << eligible.size()); ++chosen) {
		Mask row = 0;
		for (int index = 0; index < static_cast<int>(eligible.size()); ++index) {
			if (chosen & (std::uint64_t{1} << index)) {
				row |= Mask{1} << eligible[index];
			}
		}
		for (int fresh = 0; size(row) + fresh <= 7; ++fresh) {
			if (size(row) + fresh < 3) continue;
			Mask extra = row;
			for (int index = 0; index < fresh; ++index) {
				extra |= Mask{1} << (x + 1 + index);
			}
			std::vector<Mask> rows = base;
			rows.push_back(extra);
			test_candidate(rows, kernel, x + 1 + fresh, x, summary);
		}
	}
}

void enumerate_two_row_patterns(
	const std::vector<Mask> &base,
	const std::vector<Mask> &kernel,
	const std::vector<int> &base_degrees,
	int x,
	int point,
	Mask first,
	Mask second,
	Summary &summary
)
{
	if (point <= x) {
		for (int pattern = 0; pattern < 4; ++pattern) {
			int added = bool(pattern & 1) + bool(pattern & 2);
			if (base_degrees[point] + added > 4) continue;
			enumerate_two_row_patterns(
				base,
				kernel,
				base_degrees,
				x,
				point + 1,
				first | ((pattern & 1) ? Mask{1} << point : 0),
				second | ((pattern & 2) ? Mask{1} << point : 0),
				summary
			);
		}
		return;
	}

	for (int first_private = 0; size(first) + first_private <= 7;
		++first_private) {
		for (int second_private = 0; size(second) + second_private <= 7;
			++second_private) {
			int max_shared = std::min(
				7 - size(first) - first_private,
				7 - size(second) - second_private
			);
			for (int shared = 0; shared <= max_shared; ++shared) {
				if (size(first) + first_private + shared < 3) continue;
				if (size(second) + second_private + shared < 3) continue;
				int next = x + 1;
				Mask first_row = first;
				for (int index = 0; index < first_private; ++index) {
					first_row |= Mask{1} << next++;
				}
				Mask second_row = second;
				for (int index = 0; index < second_private; ++index) {
					second_row |= Mask{1} << next++;
				}
				for (int index = 0; index < shared; ++index) {
					first_row |= Mask{1} << next;
					second_row |= Mask{1} << next++;
				}
				const std::vector<int> first_points = points(first_row);
				const std::vector<int> second_points = points(second_row);
				if (std::lexicographical_compare(
					second_points.begin(),
					second_points.end(),
					first_points.begin(),
					first_points.end()
				)) continue;
				std::vector<Mask> rows = base;
				rows.push_back(first_row);
				rows.push_back(second_row);
				test_candidate(rows, kernel, next, x, summary);
			}
		}
	}
}

std::vector<std::string> read_classes(const std::string &path)
{
	std::ifstream input(path);
	if (!input) throw std::runtime_error("cannot open kernel results");
	std::vector<std::string> result;
	std::string line;
	while (std::getline(input, line)) {
		if (line.rfind("class_", 0) == 0) {
			result.push_back(line.substr(line.find('=') + 1));
		}
	}
	return result;
}

std::pair<std::vector<Mask>, int> kernel_from_key(const std::string &key)
{
	std::vector<int> supports;
	std::size_t position = 0;
	const std::size_t end = key.find('|');
	while (position < end) {
		const std::size_t colon = key.find(':', position);
		const std::size_t comma = key.find(',', colon);
		const std::size_t next = std::min(comma, end);
		int support = std::stoi(key.substr(position, colon - position));
		int count = std::stoi(key.substr(colon + 1, next - colon - 1));
		for (int copy = 0; copy < count; ++copy) supports.push_back(support);
		position = next + 1;
	}
	std::vector<Mask> rows(4);
	for (int point = 0; point < static_cast<int>(supports.size()); ++point) {
		for (int row = 0; row < 4; ++row) {
			if (supports[point] & (1 << row)) rows[row] |= Mask{1} << point;
		}
	}
	return {rows, supports.size()};
}

Summary check_class(const std::string &key)
{
	const auto [kernel, old_points] = kernel_from_key(key);
	const std::vector<Mask> kernel_blockers = blockers(kernel);
	const int x = old_points;
	std::vector<Mask> pairs;
	std::vector<Mask> others;
	for (Mask row : kernel_blockers) {
		if (size(row) == 2) pairs.push_back(row);
		else others.push_back(row);
	}
	if (pairs.size() != 3 && pairs.size() != 4) {
		throw std::runtime_error("unexpected pair-blocker count");
	}

	Summary summary;
	int first_added = -1;
	int last_added = pairs.size() == 3 ? static_cast<int>(others.size()) - 1 : -1;
	for (int added = first_added; added <= last_added; ++added) {
		std::vector<Mask> base;
		for (Mask row : pairs) base.push_back(row | (Mask{1} << x));
		for (int index = 0; index < static_cast<int>(others.size()); ++index) {
			base.push_back(others[index] | (index == added ? Mask{1} << x : 0));
		}
		int extras = 9 - base.size();
		if (extras == 0) {
			test_candidate(base, kernel, x + 1, x, summary);
		} else if (extras == 1) {
			enumerate_one_row(base, kernel, x, summary);
		} else if (extras == 2) {
			const std::vector<int> base_degrees = degrees(base, x + 1);
			enumerate_two_row_patterns(
				base, kernel, base_degrees, x, 0, 0, 0, summary
			);
		} else {
			throw std::runtime_error("unexpected kernel blocker count");
		}
	}
	return summary;
}

int main(int argc, char **argv)
{
	if (argc != 3) {
		std::cerr << "usage: middle-seven-nine-extension KERNEL_RESULTS OUTPUT\n";
		return 2;
	}
	const std::vector<std::string> classes = read_classes(argv[1]);
	if (classes.size() != 11) throw std::runtime_error("expected eleven classes");
	std::ofstream output(argv[2]);
	if (!output) {
		std::cerr << "cannot open output\n";
		return 2;
	}
	output << "schema=p0054.g4.12.middle-seven-nine-extension-cpp.v1\n";
	output << "kernel_classes=" << classes.size() << '\n';
	std::uint64_t candidates = 0;
	std::uint64_t valid = 0;
	for (int index = 0; index < static_cast<int>(classes.size()); ++index) {
		const Summary summary = check_class(classes[index]);
		output << "class_" << index + 1 << "_extension_candidates="
			<< summary.candidates << '\n';
		output << "class_" << index + 1 << "_valid_extensions="
			<< summary.valid << '\n';
		candidates += summary.candidates;
		valid += summary.valid;
	}
	output << "extension_candidates=" << candidates << '\n';
	output << "valid_extensions=" << valid << '\n';
	std::cout << "output=" << argv[2] << '\n';
	return 0;
}
